test/test_hawkmoth.py

```diff
@@ -9,7 +9,11 @@
 from hawkmoth.parser import parse_to_string
 
 def _get_output(input_filename, **options):
-    return parse_to_string(input_filename, False, **options)
+    output, errs = parse_to_string(input_filename, False, **options)
+    if errs:
+        for err in errs:
+            print(err)
```

Comment:
This is wrong, we should not print any of this here.

We need to create test cases to automatically check that the warnings and errors are generated appropriately and provide unit tests designed to produce them. I didn't do this in my example either by the way, but I believe it's needed. It's also a bit more work.

I'd hold off that work though until we settle the issue around the level of errors so that we don't waste time on half solutions.

Reply:
So warnings are like:

```
2 <SourceLocation file '/home/mvlad/src/hawkmoth/test/example-70-function.c', line 9, column 17> declaration of 'struct list' will not be visible outside of this function
```

While errors are like:
```
S 3, <SourceLocation file '/home/mvlad/src/hawkmoth/test/example-50-struct.c', line 9, column 2>, unknown type name 'list_data_t'
```

And fatal are when files are not found at all.
```
S 4, <SourceLocation file '/tmp/bla.c', line 1, column 10>, 'nofile.h' file not found
```

I haven't touched test part at all so IDK how this should behave but tests like these are quite easy to add.

Reply:
Yup, and I'm kindly requesting they go in as part of this series. :)

Rather than polluting stdout with this, we must compare with expected warnings/errors and just say: it passed/failed. We also need tests for it.

On the presentation side of things, I'm not happy with it. You're using 'Clang speak', but we will have warnings and errors that have nothing to do with Clang so we must have a generic way of showing them.

1. Severity must be about registering warnings or errors, the user has no need for some random number with a random `S` preceding it. The user must get `WARNING:` or `ERROR:` or whatever Sphinx supports.
2. File name, line and column must be presented in a Clang and Python agnostic way. You're currently dumping the `__repr__` method of a Clang object or whatever that is (I made this comment somewhere else already).

This is partly why I'd like to see some other kind of warnings and errors being raised in this series already. It would help consolidate the idea in a more general way.

Reply:
I understand that this has be improved, but in the same time, this will take much longer to do properly and it is orthogonal to current scope (imo this escalate to something else :)).

For the time being, I'm suggesting to improve testing facility in a new PR.

Reply:
If @jnikula's ok with that, so am I. The other comments still apply, but I think you addressed those at least in part.